fast64_internal/f3d/f3d_writer.py

```diff
@@ -2024,6 +2027,53 @@ def normToSigned8Vector(normal):
 	return [int.from_bytes(int(value * 127).to_bytes(1, 'big',
 		signed = True), 'big') for value in normal]
 
+def saveBitGeoGBI2(value, defaultValue, flagName, geo, matWriteMethod):
+	if value != defaultValue or matWriteMethod == GfxMatWriteMethod.WriteAll:
+		if value:
+			geo.setFlagList.append(flagName)
+		else:
+			geo.clearFlagList.append(flagName)
+
+def saveGeoModeDefinitionGBI2(fMaterial, settings, defaults, matWriteMethod):
+	geo = SPGeometryMode([],[])
+
+	saveBitGeoGBI2(settings.g_zbuffer, defaults.g_zbuffer, 'G_ZBUFFER',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_shade, defaults.g_shade, 'G_SHADE',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_cull_front, defaults.g_cull_front, 'G_CULL_FRONT',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_cull_back,  defaults.g_cull_back, 'G_CULL_BACK',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_fog, defaults.g_fog, 'G_FOG', geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_lighting, defaults.g_lighting, 'G_LIGHTING',
+		geo, matWriteMethod)
+
+	# make sure normals are saved correctly.
+	saveBitGeoGBI2(settings.g_tex_gen, defaults.g_tex_gen, 'G_TEXTURE_GEN',
+		geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_tex_gen_linear, defaults.g_tex_gen_linear,
+		'G_TEXTURE_GEN_LINEAR', geo, matWriteMethod)
+	saveBitGeoGBI2(settings.g_shade_smooth, defaults.g_shade_smooth,
+		'G_SHADING_SMOOTH', geo, matWriteMethod)
+	if bpy.context.scene.f3d_type == 'F3DEX_GBI_2' or \
+		bpy.context.scene.f3d_type == 'F3DEX_GBI':
```

Comment:
this condition is always true in this function

Reply:
Yeah I'm not actually sure why it is like that, this function is just a modified version of another function, and it was like that originally.

Reply:
well `saveGeoModeDefinition` deals with many microcodes, `saveGeoModeDefinitionGBI2` only with f3dex2

according to oot decomp's gbi.h `G_CLIPPING` only exists in f3dex/f3dex2 so that's why